proof_of_concept/definitions/assets.py

```diff
@@ -31,18 +31,19 @@ def __init__(self, job: Job, item: str) -> None:
 class Asset:
     """Asset, a representation of a computation or piece of data."""
 
-    def __init__(self, id: Union[str, Identifier], data: Any,
+    def __init__(self, id: Union[str, Identifier], kind: str, data: Any,
                  image_location: Optional[str] = None,
                  metadata: Optional[Metadata] = None
                  ) -> None:
         """Create an Asset.
 
         Assets may have either a data item (a JSON-serialisable Python
-        value) associated with them, or reference Docker image tarball
+        value) associated with them, or reference a Docker image tarball
         containing data or code.
 
         Args:
             id: Identifier of the asset
+            kind: Either "data" or "compute"
```

Comment:
Could we not instead use the type of an object (i.e. `ComputeAsset` or `DataAsset`) to know the kind of asset? Or did you already try that and it didn't work out?

Reply:
Ah I see we need this for deserialisation right?

Reply:
So far, we've distinguished assets by whether the data field was `None`, but now that we have data asset with an image rather than a value, that no longer works. So I added the attribute, and I may well decide to remove the direct execution, remove the `ComputeAsset` and `DataAsset` classes, and only have `Asset`, but with the `kind` attribute so we can check that someone didn't accidentally try to use a data asset to compute something. Not sure entirely how that will play out, but I think having it is an improvement.

Reply:
Check